File: haskell/src/Solutions/Euler172.hs
module Solutions.Euler172 where
import Utils.Numeric(binom, factorial)

u n = sum [ans |
    a <- [0..n `div` 3], -- digits appearing thrice
    b <- [0..n `div` 2], -- digits appearing twice
    c <- [0..n], -- digits appearing once
    let d = 9 - a - b - c, -- digits not appearing
    d >= 0,
    3 * a + 2 * b + c == n,
    -- choose digits appearing x times * choose their places * choose their alignment
    let ans = binom 9 a * binom n (3 * a) * factorial (3 * a) `div` (factorial 3 ^ a)
            * binom (9 - a) b * binom (n - 3 * a) (2 * b) * factorial (2 * b) `div` (factorial 2 ^ b)
            * binom (9 - a - b) c * factorial c]

euler172 :: IO String
euler172 = return $ show $ 9 * (u 17 + 17 * u 16 + 17 * 8 * u 15)

File: rust/src/solutions/euler172.rs
use crate::utils::numeric::factorial::factorial as fac;
use num::integer::binomial as bin;

fn count(n: u128) -> u128 {
    let mut ans = 0;
    // Digits appearing thrice
    for a in 0..=n / 3 {
        // Digits appearing twice
        for b in 0..=n / 2 {
            // Digits appearing once
            for c in 0..=n {
                if 3 * a + 2 * b + c == n && a + b + c <= 9 {
                    // Choose digits appearing x times * choose their alignment * choose their places
                    let n3 = bin(9, a) * bin(n, 3 * a) * fac(&(3 * a)) / fac(&3u128).pow(a as u32);
                    let n2 = bin(9 - a, b) * bin(n - 3 * a, 2 * b) * fac(&(2 * b))
                        / fac(&2u128).pow(b as u32);
                    let n1 = bin(9 - a - b, c) * fac(&c);
                    ans += n3 * n2 * n1;
                }
            }
        }
    }
    ans
}

pub fn euler172() -> String {
    (9 * (count(17) + 17 * count(16) + 17 * 8 * count(15))).to_string()
}
